### ISSUE TITLE
[BUG][V1.7.0][Longhornctl] Longhorn CLI fail on check preflight on Fedora CoreOS

### ISSUE THREAD

**Issue by Nemric** [longhorn/longhorn#9301]
## Describe the bug
``` shell
emeric@node:~$ ./longhornctl check preflight --kube-config ~/.kube/config 
INFO[2024-08-25T13:46:53+02:00] Initializing preflight checker               
INFO[2024-08-25T13:46:53+02:00] Cleaning up preflight checker                
INFO[2024-08-25T13:46:53+02:00] Running preflight checker                    
ERRO[2024-08-25T13:47:08+02:00] Failed to run preflight checker: failed DaemonSet condition check: pod container is in crash loop. View the logs using "kubectl -n default logs longhorn-preflight-checker-2p2g7 -c init-longhornctl" 
emeric@node:~$ kubectl -n default logs longhorn-preflight-checker-2p2g7 -c init-longhornctl
time="2024-08-25T11:47:22Z" level=info msg="Checking iscsid service status"
time="2024-08-25T11:47:22Z" level=info msg="Checking multipathd service status"
time="2024-08-25T11:47:23Z" level=info msg="Checking if NFS4 (either 4.0, 4.1 or 4.2) is supported"
time="2024-08-25T11:47:23Z" level=error msg="Failed to run preflight checker: open /host/boot/config-6.9.11-200.fc40.x86_64: no such file or directory"
```

So I did try to install longhorn with the kubectl apply method then, `longhorn-driver-deployer` is stuck and it looks like it's waiting for `longhorn-preflight-checker` pods that fail.
I don't really now what `longhorn-preflight-checker` does but the name let me think it's a longhornctl pod ;)

## To Reproduce

I did boot my Fedora CoreOS baremetal nodes via PXE/iPXE and provision them via ignition files.
k8s works well as far as I can tell (I'm in early steps of building a volatile kube)
I think that in this configuration, I don't have any `/boot/config-6.9.11-200.fc40.x86_64` directory so you can't rely on it

## Expected behavior

Preflight check should continue 

## Environment

 - Longhorn version: 1.7.0
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: Kubeadm 1.30.4
   - Number of control plane nodes in the cluster: 1
   - Number of worker nodes in the cluster: 3
 - Node config
   - OS type and version: Fedora CoreOS 40
   - Kernel version: 6.9.11-200.fc40
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal

**Comment by derekbit** [comment]
@Nemric 
Thanks for reporting the issue. We will improve the compatibility issue. You could fall back to environment_check.sh instead.

**Comment by Nemric** [comment]
Thanks @derekbit :)
The biggest concern is in the `longhorn-driver-deployer` installation process that seems to rely on `longhorn-preflight-checker` that seems to use longhornctl
So, how could I move to `environment_check.sh` while using `kubectl apply .... `

**Comment by derekbit** [comment]
@Nemric You can check https://longhorn.io/docs/1.7.0/deploy/install/#using-the-environment-check-script. :)

**Comment by Nemric** [comment]
Hi, I did try this script and ... I had to tweak it :
As in [this line](https://github.com/longhorn/cli/blob/ed542a0d6bf1cc96d6bd35d1278cabf3d10c2421/pkg/local/preflight/checker.go#L425) : path to kernel config is hard coded, not a problem for many distrib but for CoreOS it is
I did found this path that did the trick : `/lib/modules/${kernel}/config`

Another thing, CoreOS do not use `nfs-utils` but `nfs-utils-coreos` because of python (that is not shipped) dependency I think
``` shell
nfs-utils : NFS utilities and supporting clients and daemons for the kernel NFS server
nfs-utils-coreos : Minimal NFS utilities for supporting clients
```

I think that [when checking](https://github.com/longhorn/cli/blob/ed542a0d6bf1cc96d6bd35d1278cabf3d10c2421/pkg/utils/os.go#L26) OS you should check for VARIANT_ID in `/etc/os-release` because CoreOS is rpm-ostree based and won't really works as fedora sever

Thanks for reading :)

**Comment by derekbit** [comment]
Thanks @Nemric. The environement_check.sh doesn't support all OSes in the market. We usually improve it when there is feedback from users. We will improve the flight check CLI according to your suggestion.
Currently, you can install and validate the requirements mentioned in https://longhorn.io/docs/1.7.0/deploy/install/. Sorry for the inconvenience.

**Comment by Nemric** [comment]
:/ I don't know what was changed but the installation of v1.7.1 did run as expected
I did upgrade K8S and Cri-O to v1.31.0 and did give another try to Longhorn and everything works great now

I don't know if I have to close the issue as I can see 2 questions :
- It seems that the longhorn-preflight-checker didn't block installation any more and perhaps have an "assume-clean" default behavior
- longhorn-preflight-checker perhaps still need to support some container optimised distribs like CoreOS or Flatcar or else

**Comment by smokes2345** [comment]
I'm not sure if this is related, but i can't get a cluster to deploy on fedora core and this is the closest thing i've found to any sort of relevant google result.  it looks like the `node` resource is having trouble validating dependencies as well, specifically, it can't verify the nfs and dm_crypt kernel modules due to the nodes not having the /proc/config.gz file.  

> Failed to check kernel config file for kernel modules map[CONFIG_NFS_V4:nfs CONFIG_NFS_V4_1:nfs CONFIG_NFS_V4_2:nfs]: failed to get kernel config map from /host/proc: open /host/proc/config.gz: no such file or directory

i'm not sure there's a good way to enable this with fedora core, is there a way to disable the node checks?

**Comment by derekbit** [comment]
> I'm not sure if this is related, but i can't get a cluster to deploy on fedora core and this is the closest thing i've found to any sort of relevant google result. it looks like the `node` resource is having trouble validating dependencies as well, specifically, it can't verify the nfs and dm_crypt kernel modules due to the nodes not having the /proc/config.gz file.
> 
> > Failed to check kernel config file for kernel modules map[CONFIG_NFS_V4:nfs CONFIG_NFS_V4_1:nfs CONFIG_NFS_V4_2:nfs]: failed to get kernel config map from /host/proc: open /host/proc/config.gz: no such file or directory
> 
> i'm not sure there's a good way to enable this with fedora core, is there a way to disable the node checks?
> 
> Using longhorn 1.18 and fedora core 41

What's your Longhorn version?
The checks won't impact the Longhorn's functionalities and just emit warnings.
The upcoming v1.8.1 and v1.7.3 will only check the OSes listed in https://longhorn.io/docs/1.8.0/best-practices/#operating-system.

**Comment by smokes2345** [comment]
i updated my original comment with longhorn and fedora versions.  it looks as though longhorn *has* finished setting up on my cluster.  the only thing i changed, after the install has been stuck for more than 24hrs, was to install nfs-utils package on one of six hosts in my cluster.  nfs-utils-coreos was already installed, providing all the supporting modules and libraries.

**Comment by derekbit** [comment]
> i updated my original comment with longhorn and fedora versions. it looks as though longhorn _has_ finished setting up on my cluster. the only thing i changed, after the install has been stuck for more than 24hrs, was to install nfs-utils package on one of six hosts in my cluster. nfs-utils-coreos was already installed, providing all the supporting modules and libraries.

Yes, it is due to Longhorn doesn't handle fedora-coreos in previous versions.

**Comment by derekbit** [comment]
> while 2nd try (v1.7.1) did run as expected.

We don't have any fix in v1.7.1, so it sounds weird...

> I'm now running v1.8.0 on CoreOS 41.20250117.3.0 (latest stable from january)

Yes, v1.8.0 already fixes the issue.

**Comment by Nemric** [comment]
Yes, weird .... but facts are here : https://github.com/longhorn/longhorn/issues/9301#issuecomment-2351556760
That's why I say that's the install process is unpredictable ...
For v1.8.0 I did an upgrade, from v1.7.2 ( and from v1.7.1), not a full install (perhaps it's the same internally) 
I did use Argocd for install and upgrade with default settings except : 
```
defaultSettings.backupTarget nfs://nfs.server:/backups/longhorn-volumes
metrics.serviceMonitor.enabled false
defaultSettings.nodeDrainPolicy block-for-eviction-if-contains-last-replica
```
